tests/test_timestamp.py

```diff
@@ -0,0 +1,149 @@
+        ), 0),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, None,
+        precision=TimestampPrecision.SECOND, fractional_seconds=Decimal('0.123456')
+        ), 123456),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, None,
+        precision=TimestampPrecision.SECOND, fractional_seconds=Decimal('0.123456789')
+        ), 123456),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, None,
+        precision=TimestampPrecision.SECOND, fractional_seconds=Decimal('0.000123')
+        ), 123),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, None,
+        precision=TimestampPrecision.SECOND, fractional_seconds=Decimal('0.123000')
+        ), 123000)
+]
+
+
+MISSING_FRACTIONAL_PRECISION = [
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, 0,
+        precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+        ), 0),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, 123456,
+        precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+        ), Decimal('0.123456')),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, 123,
+        precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+        ), Decimal('0.000123')),
+    (Timestamp(
+        2011, 1, 1,
+        0, 0, 0, 123000,
+        precision=TimestampPrecision.SECOND, fractional_precision=6, fractional_seconds=None
+        ), Decimal('0.123000'))
+]
+
+
+@listify
+def event_type_parameters(list_name):
+    print(list_name)
+    for timestamp, expected_val in list_name:
+        yield _P(
+            timestamps=timestamp,
+            expected_value=expected_val,
+        )
+
+
+@parametrize(*event_type_parameters(MISSING_MICROSECOND))
+def test_missing_microsecond(item):
+    timestamp = item.timestamps
+    expected_microsecond = item.expected_value
+    assert timestamp.microsecond == expected_microsecond
+
+
+@parametrize(*event_type_parameters(MISSING_FRACTIONAL_PRECISION))
+def test_missing_fractional_seconds(item):
+    timestamp = item.timestamps
+    expected_fractional_second = item.expected_value
+    assert timestamp.fractional_seconds == expected_fractional_second
+
+
+def test_fractional_precision_with_no_microseconds():
+    with pytest.raises(ValueError):
+        Timestamp(
+            2011, 1, 1,
+            0, 0, 0, None,
+            precision=TimestampPrecision.SECOND, fractional_precision=6
+        )
+
+
+def test_fractional_precision_less_than_1():
+    with pytest.raises(ValueError):
+        Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 0,
+            precision=TimestampPrecision.SECOND, fractional_precision=0, fractional_seconds=0
+        )
+
+
+def test_fractional_seconds_greater_than_1():
+    with pytest.raises(ValueError):
+        Timestamp(
+            2011, 1, 1,
+            0, 0, 0, 0,
+            precision=TimestampPrecision.SECOND, fractional_precision=1, fractional_seconds=2
```

Comment:
Wouldn't this also cause an error because both `fractional_seconds` and `fractional_precision` are specified, and we've made them mutually exclusive?

Reply:
Correct. Changed this test so that `fractional_precision` is `None` and `fractional_seconds` has a value.